### PR TITLE
Bugfix/middleware handlers

### LINKED ISSUE #281: Retry handler decreases the option's retry attempts
Follow up to #258
(we should really inspect all middleware handlers)

This means if multiple requests need to retry within the lifetime of the client/request adapter/handler, the first requests will retry successfully but the following ones won't. (same issue as redirect handler)

https://github.com/microsoft/kiota-http-python/blob/f64b0f8d478fae3da5edf7833041c92f9d91fc4c/kiota_http/middleware/retry_handler.py#L97

### LINKED ISSUE #299: Redirect handling does not close old Response, leading to resources not being closed
**Environment**

- Python Version: 3.11.5
- kiota-http version: 1.3.0
- OS: Linux (openSuse)

**Stack trace (if available)**

**Describe the bug**
The RedirectHandler class in the Microsoft Kiota HTTP Python library is currently experiencing an issue where the original response, containing the new URL, is not closed before performing a redirect. This behaviour results in httpcore async layer running out of streams, reaching a limit of 100 streams.

**To Reproduce**

1. Execute a request using the RedirectHandler in a scenario that triggers a redirect. I used the ms graph sdk with a drive item content.get () call to download multiple files (over 100). 
2. Observe the httpcore async behaviour after a redirect, after 100, the call to content.get() blocks (if using ms graph sdk). 

**Expected behavior**
I think the original response should be closed properly before the redirect is performed to prevent httpcore async from running out of streams.

**Screenshots**

**Additional context**
To "fix" the issue, I have used the following code, not sure if this is the correct solution. 

```python
            if redirect_location and current_options.should_redirect:
                max_redirect -= 1
                if not self.increment(response, max_redirect, history[:]):
                    break
                _redirect_span.set_attribute(REDIRECT_COUNT_KEY, len(history))
                new_request = self._build_redirect_request(request, response)
                request = new_request
                # make sure the response is closed
                await response.aclose()
                continue

            response.history = history
            break
```

### PR DESCRIPTION
**Pull request by samwelkanda**
## Overview

- Bugfix issues with middleware maintaining state across requests.[
- Fix issue with redirect handler not closing old responses.

- ## Related Issue

Fixes #281 
Fixes #299

### CODE REVIEW

**Review by andrueastman — APPROVED** [review]

File: kiota_http/middleware/redirect_handler.py [hunk 1]
```diff
@@ -168,6 +174,13 @@ def _redirect_url(self, request: httpx.Request, response: httpx.Response) -> htt
         except Exception as exc:
             raise Exception(f"Invalid URL in location header: {exc}.")
 
+        if url.scheme != request.url.scheme and not options.allow_redirect_on_scheme_change:
```
Comment (baywet) [inline kiota_http/middleware/redirect_handler.py:177]: Maybe consider allowing an upgrade from http to https? It sometimes happens with CDNs, and some people publish static generated APIs

**Review by baywet — APPROVED** [review]

File: kiota_http/middleware/redirect_handler.py (hunk 1 above)
Comment (samwelkanda) [reply kiota_http/middleware/redirect_handler.py:177]: Makes sense. Should we make this the default behavior across the http packages? For instance dotnet doesn't support that: https://github.com/microsoft/kiota-http-dotnet/blob/119e79cf50dcb94770ba3942ff13e1e9ab57bf8e/src/Middleware/RedirectHandler.cs#L130C1-L131C1

Comment (baywet) [reply kiota_http/middleware/redirect_handler.py:177]: I don't have all the context for that original implantation but maybe @andrueastman does ?
Other than that I'd say yes.

Comment (andrueastman) [reply kiota_http/middleware/redirect_handler.py:177]: The origin of this was to initially prevent downgrading to a less secure protocol (e.g https to http) for security reasons as the goal was to be secure by default and let the user opt in to sending/receiving data to/from a potentially unsecure endpoint. https://github.com/microsoft/kiota/pull/589#discussion_r703752506

Given that this behavior is still technically configurable through the `ShouldRedirect` lambda(can technically be used to whitelist schemes) and the `allow_redirect_on_scheme_change`, I don't think we need to change anything here for now and probably handle the discussion separately as changing this now will probably introduce code that checks and compares secure and insecure schemes that are valid for upgrades(unless http to https is the only scenario here?).

Comment (baywet) [reply kiota_http/middleware/redirect_handler.py:177]: Thanks for the additional information!
http to https only scenario: yes as we depend on an http client under the hood anyway. Some "http clients" used to be able to do FTP/SFTP or SMB, but I think most of those have disappeared in favor of protocol specific clients over time.

Handling this in a different issue to unblock this PR: sure! @samwelkanda can you create an issue in the sdk design repo to update the spec so it includes that aspect please?

**Pull request merged**